### ISSUE TITLE
iOS keyboard support

### ISSUE THREAD

**Issue by ThatsNiceGuy** [ppy/osu#4777]
I am new to github and reporting stuff like this here but I hope I did right lol
**Describe the new feature:** 
Enable iOS users to use a physical keyboard for gameplay, like a Smart Keyboard on iPad Pro, or if possible maybe even a Bluetooth keyboard because right now you can only use a keyboard for text entry fields, like the chat or any text box like the one in osu!direct.
(Testing on an iPad Pro 10.5” 2017 with a Smart Keyboard)

**Proposal designs of the feature:** 
Basically let us use a physical keyboard in iOS for gameplay like Autopilot in STD, taiko, or CtB.

**Comment by jai-x** [comment]
It seems that keyboard support currently is very partial as some people report Smart Keyboard and other keyboards working with iOS while others report it as broken. See this issue thread: https://github.com/ppy/osu/issues/4225

I don't know why that issue was closed, but keep this one open. Will most likely require a framework level fix.

**Comment by ThatsNiceGuy** [comment]
That’s kinda weird how it works for some but not others. Well alright I’ll leave this open

**Comment by MiraiSubject** [comment]
Raw input for the keyboard is merged ([osu-framework/2087](https://github.com/ppy/osu-framework/pull/2087)) and I got it working exactly once when I attempted to race the textfield trigger with opening a beatmap which caused the game to accept keyboard input for standard gameplay, because iOS still thought a textfield was active.

If you open the search field in the beatmap selector and play around with the up and down arrow keys you should be able to cycle through your beatmaps which didn't work before that PR was merged, so  raw input does work.

(The referenced #4225 issue was about external keyboard support for typing, not necessarily limited to gameplay itself. )

**Comment by ThatsNiceGuy** [comment]
so it's like an iOS limitation that we can't use keyboards without a textfield being open?
I did try the search field thing and it did work for me which seems like a good sign.
I wonder what it would take to get that functionality but everywhere in the game or just in gameplay.

**Comment by jai-x** [comment]
From the framework fix linked, it appears that the osu! side code wasn't changed to take advantage of it. Specifically this line:
https://github.com/ppy/osu/blob/ce5bd71f2db32d7e812d462b503f5ea09789b062/osu.iOS/Application.cs#L12

Should be changed to:
```csharp
UIApplication.Main(args, "GameUIApplication", "AppDelegate");
```

Just double checking, is that correct? @imxieyi

**Comment by imxieyi** [comment]
> From the framework fix linked, it appears that the osu! side code wasn't changed to take advantage of it. Specifically this line:
> https://github.com/ppy/osu/blob/ce5bd71f2db32d7e812d462b503f5ea09789b062/osu.iOS/Application.cs#L12
> 
> Should be changed to:
> 
> ```cs
> UIApplication.Main(args, "GameUIApplication", "AppDelegate");
> ```
> 
> Just double checking, is that correct? @imxieyi

Yes. I personally tried this a few days ago. It should work naturally by just modifying this line of code.

**Pay attention that this feature relies heavily on a private API that has been existing in iOS from at least iOS 6.0. However, there is no other known ways to achieve raw input. There are actually a few apps on the app store that utilize this.**

**Comment by ThatsNiceGuy** [comment]
so if theres other apps on the app store that use this same api then maybe lazer will eventually reach the app store? damn im just waiting for what's next for the future of this game

**Comment by imxieyi** [comment]
> so if theres other apps on the app store that use this same api then maybe lazer will eventually reach the app store? damn im just waiting for what's next for the future of this game

I didn't dig into them. But at least [VNC Viewer](https://itunes.apple.com/us/app/vnc-viewer-remote-desktop/id352019548?mt=8) and [nPlayer](https://itunes.apple.com/cn/app/nplayer/id1116905928?mt=8) can capture raw keyboard input. This is not achievable using just public APIs.

**Comment by ThatsNiceGuy** [comment]
alright thats interesting

**Comment by smoogipoo** [comment]
Fixed by referencing PR.
